base/math.jl

```diff
@@ -867,29 +867,35 @@ end
     z
 end
 @inline function ^(x::Float32, y::Float32)
-    z = ccall("llvm.pow.f32", llvmcall, Float32, (Float32, Float32), x, y)
+    z = Float32(exp2_fast(log2(Float64(x))*y))
+    if isnan(z) & !isnan(x+y)
+        throw_exp_domainerror(x)
+    end
+    z
+end
+@inline function ^(x::Float16, y::Float16)
+    z = Float16(exp2_fast(log2(Float32(x))*y))
     if isnan(z) & !isnan(x+y)
         throw_exp_domainerror(x)
     end
     z
 end
-@inline ^(x::Float16, y::Float16) = Float16(Float32(x)^Float32(y))  # TODO: optimize
 
 @inline function ^(x::Float64, y::Integer)
     y == -1 && return inv(x)
     y == 0 && return one(x)
     y == 1 && return x
     y == 2 && return x*x
     y == 3 && return x*x*x
-    ccall("llvm.pow.f64", llvmcall, Float64, (Float64, Float64), x, Float64(y))
+    return x^Float64(y)
```

Comment:
Why weren't we using the llvm intrinsic for this (isnan(z) should be impossible, IIUC, so the error checking in the call to `^` isn't necessary)?
```suggestion
    ccall("llvm.powi.f64", llvmcall, Float64, (Float64, Int32), x, clamp(y, Int32))
```

Reply:
`prevfloat(1.0)^(2^33)==0.9999990463261383`, so this isn't a valid definition for large `y`

Reply:
That is a good reason. Apparently "under-the-hood" it also just uses essentially the "power-by-squaring" method, which seems to have somewhat low accuracy there:
```
julia> Base.power_by_squaring(prevfloat(1.0), 2^62)
4.3774743599864475e-223

julia> ^(prevfloat(1.0), 2^62)
4.377491037052927e-223

julia> ^(big(prevfloat(1.0)), 2^62)
4.377491037052927038343183444930184380098189891120230882184167270887044427471645e-223
```

Reply:
Yeah. Power by squaring necessarily rounds `log_2(y)` roundings